ksqldb-engine/src/main/java/io/confluent/ksql/execution/ExecutionPlanBuilder.java

```diff
@@ -87,4 +97,142 @@ public ExecutionPlan buildPhysicalPlan(
         resultStream.getSourceStep()
     );
   }
+
+
+  private void optimizeForSelfJoin(final PlanNode currentNode) {
+
+    if (currentNode instanceof JoinNode) {
+      final JoinNode joinNode = (JoinNode) currentNode;
+      if (! matchSubGraphPattern1(joinNode)) {
+        if (!matchSubGraphPattern2(joinNode)) {
+          matchSubGraphPattern3(joinNode);
+        }
+      }
+    }
+
+    for (final PlanNode child: currentNode.getSources()) {
+      optimizeForSelfJoin(child);
```

Comment:
Given the discussion from the AK PR, do you have a test case where there are two self-joins which are joined?  (E.g. where there are two joins which are optimized?)